### ISSUE TITLE
Inconsistent requirements (bytes/unicode) on argument of get_transport_and_path

### ISSUE THREAD

**Issue by inducer** [jelmer/dulwich#367]
This is in Py3. If it's `bytes`, then this line is unhappy:

https://github.com/jelmer/dulwich/blob/master/dulwich/client.py#L1220

If it's `unicode`, this line is unhappy:

https://github.com/jelmer/dulwich/blob/master/dulwich/client.py#L634

FWIW, the location string that causes this is `git://github.com/inducer/relate-sample'.

**Comment by jelmer** [comment]
Where is this causing a crash?

You indeed need to do the encoding (str -> bytes) of the path yourself at the moment, as Dulwich doesn't have a way of knowing the remote encoding.

**Comment by jelmer** [comment]
Actually, it seems like the problem is host. I've committed a fix to master that I think should address this. Please reopen if you can still reproduce it.

**Comment by inducer** [comment]
I don't have a button to reopen this issue, but it is not resolved. Let me rephrase this as a simple question--is the `location` argument to `get_transport_and_path` `str` or `bytes` in Py3? The documentation should state this, but right now it does not (as far as I can tell).

You seem to indicate it's bytes, but that fails right away, in the same function:

https://github.com/jelmer/dulwich/blob/master/dulwich/client.py#L1240

for this reason:

```
>>> '@' in b'stuff' 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'str' does not support the buffer interface
```

Another thing that indicates `str` is likely the right way to go is that you're going to be opening a bunch of files at a path obtained from this function using Py3 file I/O, and that expects Unicode file names.

Based on your comment that it should be `bytes`, I went down a rabbit hole of trying to clean that up, the wreckage of which is here:

https://gist.github.com/inducer/1cd1afe168a950b8dfce

FWIW, things seem to work OK if I pass an SSH URL `git@bitbucket.org:something/oranother.git` as Unicode.

On the other hand, if I pass a `git://somewhere/something.git` URL, that fails here:

https://github.com/jelmer/dulwich/blob/36d40a4d41a8f7a81ab5f4e1f30488935a3cbb6b/dulwich/client.py#L610

**Comment by jelmer** [comment]
The location argument to get_transport_and_path() should be a *str* in Python 3. That will return the path as a string, which you need to encode if you're going to be using TcpGitClient, which takes bytes. I've updated the docstrings.

**Comment by inducer** [comment]
(meant to comment here) Thank you!
